### PR TITLE
fix(vscode): open external links in preview panel via extension host

### LINKED ISSUE #2422: In Extension Diagram Preview URLs/Links do not open target
### Relates to

VSCode extension

### Version you are using

1.45.0

### Observation:
In the VSCode Extension Diagram Preview the click on URLs defined in a element markdown description or via "link" keyword does not open an internal/external browser window. Instead the Diagram window shows a blank page.
In the linked Playground example running in the browser it works:
Image

### Expectation:
Click on link either opens the web page in VSCode or in an external browser

### If possible, include link to the playground with example

https://playground.likec4.dev/w/07YOBx/index/

### PR DESCRIPTION
**Pull request by ckeller42**
## Summary

Fix for #2422 — clicking URLs in the VSCode extension diagram preview showed a blank page instead of opening the browser.

### Root Cause

VSCode webviews cannot open URLs directly via `<a href="..." target="_blank">`. External link navigation requires routing through the extension host via `vscode.env.openExternal()`.

### Fix

1. **Extracted `interceptExternalLinks`** — testable function that catches `click` + `auxclick` events on `http(s)` anchors, with `instanceof Element` guard and `closest()` for nested elements
2. **New `OpenExternalUrl` notification** in the webview↔extension protocol
3. **Host-side handler** with defense-in-depth: `vscode.Uri.parse(url, true)` strict parsing + scheme validation (http/https only) + error logging
4. **HMR cleanup** via `import.meta.hot.dispose()` to prevent double-registration in dev mode

### Tests (13 unit tests)

- https/http interception with `preventDefault`
- Case-insensitive scheme matching (`HTTPS://`)
- Nested element clicks (`closest('a[href]')` bubble-up)
- `auxclick` (middle-click) interception
- Security: `javascript:`, `data:`, `file:`, `mailto:` rejection
- Edge cases: null targets, non-elements, missing href, cleanup
- Wiring test: both listeners registered and cleanup removes them

### Files changed

- `packages/vscode-preview/src/interceptExternalLinks.ts` — extracted handler
- `packages/vscode-preview/src/interceptExternalLinks.spec.ts` — 13 unit tests
- `packages/vscode-preview/src/main.tsx` — wiring with HMR cleanup
- `packages/vscode-preview/src/vscode.ts` — `openExternalUrl()` in ExtensionApi
- `packages/vscode-preview/protocol.ts` — `OpenExternalUrl` notification type
- `packages/vscode-preview/vitest.config.ts` — test infrastructure (follows `defineVitest` pattern)
- `packages/vscode-preview/package.json` — added `vitest`, `test` script
- `packages/vscode/src/useMessenger.ts` — handler registration
- `packages/vscode/src/panel/activateMessenger.ts` — `vscode.env.openExternal()` with scheme validation

Closes #2422

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by ckeller42** [comment]
@coderabbitai review

File: packages/vscode-preview/src/main.tsx [hunk 1]
```diff
@@ -7,6 +7,19 @@ import { IconsProvider } from './IconRenderer'
...
+// Intercept external link clicks and route them through the extension host,
+// since VSCode webviews cannot open URLs directly (#2422).
+document.addEventListener('click', (e) => {
+  const anchor = (e.target as HTMLElement).closest('a[href]')
+  if (!anchor) return
+  const href = anchor.getAttribute('href')
+  if (href && /^https?:\/\//.test(href)) {
+    e.preventDefault()
+    ExtensionApi.openExternalUrl(href)
+  }
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply packages/vscode-preview/src/main.tsx]: Fixed — replaced `as HTMLElement` with `instanceof Element` guard and added case-insensitive regex.

File: packages/vscode/src/panel/activateMessenger.ts [hunk 2]
```diff
@@ -107,6 +107,10 @@ export function activateMessenger() {
     await executeCommand(commands.locate, params)
   })
 
+  messenger.onWebviewOpenExternalUrl(({ url }) => {
+    vscode.env.openExternal(vscode.Uri.parse(url))
+  })
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply packages/vscode/src/panel/activateMessenger.ts]: Fixed — added scheme validation (http/https only), await, and error logging.

**Comment by ckeller42** [comment]
@coderabbitai review

**Comment by ckeller42** [comment]
@coderabbitai review

**Comment by ckeller42** [comment]
@coderabbitai review

**Comment by ckeller42** [comment]
@coderabbitai review

**Comment by ckeller42** [comment]
@coderabbitai review

File: packages/vscode-preview/src/interceptExternalLinks.spec.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  it('does nothing when target is null', () => {
+    handleExternalLinkClick(mockEvent(null), openUrl)
+    expect(openUrl).not.toHaveBeenCalled()
+  })
+
+  it('does nothing when closest returns null (no parent anchor)', () => {
+    handleExternalLinkClick(mockEvent(mockAnchor(null)), openUrl)
+    expect(openUrl).not.toHaveBeenCalled()
+  })
+})
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply packages/vscode-preview/src/interceptExternalLinks.spec.ts:85]: Already addressed — the latest commit includes a `describe('interceptExternalLinks wiring')` test that verifies both `click` and `auxclick` listeners are registered and cleanup removes them.

File: packages/vscode-preview/src/interceptExternalLinks.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+const isHttpUrl = /^https?:\/\//i
+
+export function handleExternalLinkClick(
+  e: MouseEvent,
+  openExternalUrl: (url: string) => void,
+): void {
+  const target = e.target
+  if (!target || typeof (target as Element).closest !== 'function') return
+  const anchor = (target as Element).closest('a[href]')
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply packages/vscode-preview/src/interceptExternalLinks.ts:18]: Intentional trade-off: we use duck-typing (`typeof target.closest !== 'function'`) instead of `instanceof Element` because the unit tests run in Node.js without jsdom. The `as Element` cast is needed for TypeScript after the duck-type check. The `as MouseEvent` cast on the handler is necessary at the `addEventListener` boundary (which types events as `Event`, not `MouseEvent`).

**Comment by ckeller42** [comment]
@coderabbitai review

**Review by davydkov — APPROVED** [review]
